File: CS/ReportingEUDSaveReportWithoutConnectionParams/CustomConnectionStorageService.cs
using DevExpress.DataAccess.ConnectionParameters;
using DevExpress.DataAccess.Sql;
using DevExpress.DataAccess.Wizard.Model;
using DevExpress.DataAccess.Wizard.Native;
using DevExpress.DataAccess.Wizard.Services;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using System.Xml;

namespace WinEUDSaveReportWithoutConnectionParams {
    class CustomConnectionStorageService : IConnectionStorageService {
        public bool CanSaveConnection {
            get { return true; }
        }
        
        bool IConnectionStorageService.Contains(string connectionName) {            
            return IncludeApplicationConnections ?
                DefaultStorage.Contains(connectionName) || ConnectionProviderService.GetConnectionsFromXml().Any(c => string.Equals(c.Name, connectionName)) :
                ConnectionProviderService.GetConnectionsFromXml().Any(c => string.Equals(c.Name, connectionName));
        }
        IEnumerable<DevExpress.DataAccess.Sql.SqlDataConnection> IConnectionStorageService.GetConnections() {           
            return IncludeApplicationConnections 
                ? DefaultStorage.GetConnections().Union(ConnectionProviderService.GetConnectionsFromXml()) 
                : ConnectionProviderService.GetConnectionsFromXml();
        }
        void CreateDefaultConnection() {
            if (!File.Exists(FileName)) {
                try {
                    XmlDocument doc = new XmlDocument();
                    XmlElement root = doc.CreateElement(xmlRootName);
                    doc.AppendChild(root);
                    doc.Save(FileName);
                    SqlDataConnection defaultConnection = new CustomSqlDataConnection("Default Connection", new MsSqlConnectionParameters("localhost", "NORTHWND", "", "", MsSqlAuthorizationType.Windows));
                    SaveConnection(defaultConnection.Name, defaultConnection, true);
                } catch (Exception ex) {
                    MessageBox.Show(string.Format("Cannot create '{0}' file because of exception:{1}{1}{2}", FileName,
                        Environment.NewLine, ex.Message));                    
                }
            }
        }
        public void SaveConnection(string connectionName, IDataConnection connection, bool saveCredentials) {
            connection.StoreConnectionNameOnly = true;
            try {
                XmlDocument doc = new XmlDocument();
                XmlElement root = null;
                if (File.Exists(FileName)) {
                    doc.Load(FileName);
                    root = doc.DocumentElement;
                    if (root != null) {
                        if (root.Name != xmlRootName) {
                            MessageBox.Show(string.Format("Document element is '{0}', '{1}' expected",
                                root.Name, xmlRootName));
                            return;
                        }
                        if (root.SelectSingleNode(string.Format("Connection[Name = '{0}']", connectionName)) != null)
                            return;
                    }
                }
                if (root == null) {
                    root = doc.CreateElement(xmlRootName);
                    doc.AppendChild(root);
                }

                XmlElement nameElement = doc.CreateElement("Name");
                nameElement.AppendChild(doc.CreateTextNode(connectionName));

                XmlElement connectionStringElement = doc.CreateElement("ConnectionString");
                connectionStringElement.AppendChild(doc.CreateTextNode(connection.CreateConnectionString(!saveCredentials)));

                XmlElement connectionElement = doc.CreateElement("Connection");
                connectionElement.AppendChild(nameElement);
                connectionElement.AppendChild(connectionStringElement);

                root.AppendChild(connectionElement);
                doc.Save(FileName);
            } catch (Exception ex) {
                MessageBox.Show(string.Format("Cannot save connection to '{0}' because of exception:{1}{1}{2}", FileName,
                    Environment.NewLine, ex.Message));
            }
        }
        const string xmlRootName = "Connections";

        public CustomConnectionStorageService(CustomConnectionProviderService connectionProviderService) {
            CreateDefaultConnection();
            ConnectionProviderService = connectionProviderService;
        }
        public CustomConnectionProviderService ConnectionProviderService { get; }
        public string FileName { get; set; } = "connections.xml";
        public bool IncludeApplicationConnections { get; set; } = false;
        protected ConnectionStorageService DefaultStorage { get; } = new ConnectionStorageService();
    }
}

File: VB/ReportingEUDSaveReportWithoutConnectionParams/CustomConnectionStorageService.vb
Imports DevExpress.DataAccess.ConnectionParameters
Imports DevExpress.DataAccess.Sql
Imports DevExpress.DataAccess.Wizard.Model
Imports DevExpress.DataAccess.Wizard.Native
Imports DevExpress.DataAccess.Wizard.Services
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports System.Xml

Namespace WinEUDSaveReportWithoutConnectionParams

    Friend Class CustomConnectionStorageService
        Implements IConnectionStorageService

        Public ReadOnly Property CanSaveConnection As Boolean Implements IConnectionStorageService.CanSaveConnection
            Get
                Return True
            End Get
        End Property

        Private Function Contains(ByVal connectionName As String) As Boolean Implements IConnectionStorageService.Contains
            Return If(IncludeApplicationConnections, DefaultStorage.Contains(connectionName) OrElse ConnectionProviderService.GetConnectionsFromXml().Any(Function(c) String.Equals(c.Name, connectionName)), ConnectionProviderService.GetConnectionsFromXml().Any(Function(c) String.Equals(c.Name, connectionName)))
        End Function

        Private Function GetConnections() As IEnumerable(Of SqlDataConnection) Implements IConnectionStorageService.GetConnections
            Return If(IncludeApplicationConnections, DefaultStorage.GetConnections().Union(ConnectionProviderService.GetConnectionsFromXml()), ConnectionProviderService.GetConnectionsFromXml())
        End Function

        Private Sub CreateDefaultConnection()
            If Not File.Exists(FileName) Then
                Try
                    Dim doc As XmlDocument = New XmlDocument()
                    Dim root As XmlElement = doc.CreateElement(xmlRootName)
                    doc.AppendChild(root)
                    doc.Save(FileName)
                    Dim defaultConnection As SqlDataConnection = New CustomSqlDataConnection("Default Connection", New MsSqlConnectionParameters("localhost", "NORTHWND", "", "", MsSqlAuthorizationType.Windows))
                    SaveConnection(defaultConnection.Name, defaultConnection, True)
                Catch ex As Exception
                    MessageBox.Show(String.Format("Cannot create '{0}' file because of exception:{1}{1}{2}", FileName, Environment.NewLine, ex.Message))
                End Try
            End If
        End Sub

        Public Sub SaveConnection(ByVal connectionName As String, ByVal connection As IDataConnection, ByVal saveCredentials As Boolean) Implements IConnectionStorageService.SaveConnection
            connection.StoreConnectionNameOnly = True
            Try
                Dim doc As XmlDocument = New XmlDocument()
                Dim root As XmlElement = Nothing
                If File.Exists(FileName) Then
                    doc.Load(FileName)
                    root = doc.DocumentElement
                    If root IsNot Nothing Then
                        If Not Equals(root.Name, xmlRootName) Then
                            MessageBox.Show(String.Format("Document element is '{0}', '{1}' expected", root.Name, xmlRootName))
                            Return
                        End If

                        If root.SelectSingleNode(String.Format("Connection[Name = '{0}']", connectionName)) IsNot Nothing Then Return
                    End If
                End If

                If root Is Nothing Then
                    root = doc.CreateElement(xmlRootName)
                    doc.AppendChild(root)
                End If

                Dim nameElement As XmlElement = doc.CreateElement("Name")
                nameElement.AppendChild(doc.CreateTextNode(connectionName))
                Dim connectionStringElement As XmlElement = doc.CreateElement("ConnectionString")
                connectionStringElement.AppendChild(doc.CreateTextNode(connection.CreateConnectionString(Not saveCredentials)))
                Dim connectionElement As XmlElement = doc.CreateElement("Connection")
                connectionElement.AppendChild(nameElement)
                connectionElement.AppendChild(connectionStringElement)
                root.AppendChild(connectionElement)
                doc.Save(FileName)
            Catch ex As Exception
                MessageBox.Show(String.Format("Cannot save connection to '{0}' because of exception:{1}{1}{2}", FileName, Environment.NewLine, ex.Message))
            End Try
        End Sub

        Const xmlRootName As String = "Connections"

        Public Sub New(ByVal connectionProviderService As CustomConnectionProviderService)
            CreateDefaultConnection()
            Me.ConnectionProviderService = connectionProviderService
        End Sub

        Public ReadOnly Property ConnectionProviderService As CustomConnectionProviderService

        Public Property FileName As String = "connections.xml"

        Public Property IncludeApplicationConnections As Boolean = False

        Protected ReadOnly Property DefaultStorage As ConnectionStorageService = New ConnectionStorageService()
    End Class
End Namespace
